File: api/backend/ai/reasoning_preference.py
"""Provider-neutral reasoning preference vocabulary and capability resolution."""
from __future__ import annotations
from typing import Any

PREFERENCES = ("minimum", "default", "off", "on", "minimal", "low", "medium", "high", "xhigh", "max", "ultra")
EFFORTS = ("minimal", "low", "medium", "high", "xhigh", "max", "ultra")
_PREFERENCE_SET = set(PREFERENCES)
_EFFORT_SET = set(EFFORTS)


def normalize_reasoning_preference(value: Any, fallback: str = "minimum") -> str:
    if value is True:
        return "on"
    if value is False:
        return "off"
    raw = str(value or "").strip().lower()
    if raw in {"auto", "provider", "provider_default"}:
        return "default"
    if raw in {"lowest", "lowest_available", "min"}:
        return "minimum"
    if raw == "none":
        return "off"
    if raw in _PREFERENCE_SET:
        return raw
    return fallback if fallback in _PREFERENCE_SET else "minimum"


def concrete_reasoning_preferences(reasoning: Any) -> tuple[str, ...]:
    """Concrete model options ordered from lowest to highest reasoning."""
    cap = reasoning if isinstance(reasoning, dict) else {}
    if cap.get("supported") is not True:
        return ()
    mandatory = cap.get("mandatory") is True
    control = str(cap.get("control") or "provider")
    supported = {
        str(value).strip().lower() for value in cap.get("supported_efforts", [])
        if isinstance(value, str)
    }
    out: list[str] = []
    # An explicit mandatory=false is provider evidence that reasoning is optional.
    # Some catalogues publish the effort ladder separately and transiently omit
    # can_disable/none; do not let that stale metadata erase the user's Off intent.
    if not mandatory and (cap.get("mandatory") is False or cap.get("can_disable") is True
                          or control in {"toggle", "boolean"} or "none" in supported):
        out.append("off")
    if control in {"toggle", "boolean"}:
        out.append("on")
        return tuple(out)
    if control == "levels":
        out.extend(effort for effort in EFFORTS if effort in supported)
    return tuple(out)


def minimum_reasoning_preference(reasoning: Any) -> str:
    """Lowest available is the first concrete capability option, not a fixed effort."""
    options = concrete_reasoning_preferences(reasoning)
    return options[0] if options else "default"


def supported_reasoning_preferences(reasoning: Any) -> tuple[str, ...]:
    cap = reasoning if isinstance(reasoning, dict) else {}
    if not isinstance(cap.get("supported"), bool):
        return ("default",)
    if cap.get("supported") is False:
        return ()
    concrete = concrete_reasoning_preferences(cap)
    out: list[str] = []
    if concrete:
        out.append("minimum")
    out.append("default")
    out.extend(concrete)
    return tuple(out)


def resolve_reasoning_preference(value: Any, reasoning: Any) -> str:
    requested = normalize_reasoning_preference(value, "minimum")
    cap = reasoning if isinstance(reasoning, dict) else {}
    minimum = minimum_reasoning_preference(cap)
    if requested == "minimum":
        return minimum
    # Off is persistent user intent.  Capability discovery may be stale or may
    # describe a model that simply has no reasoning control; neither case may
    # silently turn the saved selection into Low/Default.  Only an explicitly
    # mandatory-reasoning model is allowed to clamp Off at dispatch time.
    if requested == "off":
        if cap.get("mandatory") is True:
            return minimum if minimum != "default" else "default"
        return "off"
    supported = supported_reasoning_preferences(cap)
    if not supported:
        return "default"
    if requested in supported:
        return requested
    return "default"


def reasoning_is_active(value: Any, reasoning: Any) -> bool:
    cap = reasoning if isinstance(reasoning, dict) else {}
    selected = resolve_reasoning_preference(value, cap)
    if selected == "off":
        return False
    if selected == "on" or selected in _EFFORT_SET:
        return True
    return cap.get("mandatory") is True or cap.get("default_enabled") is True

File: src/shared/reasoning-preference.js
/** Provider-neutral user intent for model reasoning/thinking controls. */
export const REASONING_PREFERENCES = Object.freeze([
  "minimum", "default", "off", "on", "minimal", "low", "medium", "high", "xhigh", "max", "ultra",
]);

const PREFERENCE_SET = new Set(REASONING_PREFERENCES);
const EFFORTS = Object.freeze(["minimal", "low", "medium", "high", "xhigh", "max", "ultra"]);
const EFFORT_SET = new Set(EFFORTS);

export function normalizeReasoningPreference(value, fallback = "minimum") {
  if (value === true) return "on";
  if (value === false) return "off";
  const raw = String(value ?? "").trim().toLowerCase();
  if (raw === "auto" || raw === "provider" || raw === "provider_default") return "default";
  if (raw === "lowest" || raw === "lowest_available" || raw === "min") return "minimum";
  if (raw === "none") return "off";
  if (PREFERENCE_SET.has(raw)) return raw;
  const safeFallback = String(fallback || "minimum").trim().toLowerCase();
  return PREFERENCE_SET.has(safeFallback) ? safeFallback : "minimum";
}

export function normalizeUserReasoningPreference(value) {
  // Persist the policy, not its current model-specific result. New/reset
  // profiles choose Lowest available; explicit Off/Low/etc. remain untouched.
  // The leaf adapter resolves this intent only when building a request.
  const normalized = normalizeReasoningPreference(value, "minimum");
  // Provider default remains an internal transport fallback only. Older saved
  // Provider-default/Auto choices migrate to the user-facing Lowest available intent.
  return normalized === "default" ? "minimum" : normalized;
}

export function reasoningEffortPreference(value) {
  const normalized = normalizeReasoningPreference(value, "default");
  return EFFORT_SET.has(normalized) ? normalized : "";
}

/** Return the concrete model options in ascending reasoning cost/strength. */
export function concreteReasoningPreferences(reasoning) {
  const cap = reasoning && typeof reasoning === "object" ? reasoning : {};
  if (cap.supported !== true) return [];
  const mandatory = cap.mandatory === true;
  const control = String(cap.control || "provider");
  const supported = new Set(Array.isArray(cap.supported_efforts)
    ? cap.supported_efforts.map(value => String(value || "").trim().toLowerCase())
    : []);
  const options = [];
  // Explicit mandatory=false is enough to preserve Off as an executable user
  // intent even when a stale catalogue snapshot omits can_disable/none.
  if (!mandatory && (cap.mandatory === false || cap.can_disable === true || control === "toggle" || control === "boolean" || supported.has("none")))
    options.push("off");
  if (control === "toggle" || control === "boolean") {
    options.push("on");
    return options;
  }
  if (control === "levels")
    for (const effort of EFFORTS) if (supported.has(effort)) options.push(effort);
  return options;
}

/** Lowest available is an alias for the first concrete option, never a fixed effort. */
export function minimumReasoningPreference(reasoning) {
  return concreteReasoningPreferences(reasoning)[0] || "default";
}

export function reasoningOptionsForCapability(reasoning) {
  const cap = reasoning && typeof reasoning === "object" ? reasoning : null;
  if (!cap || typeof cap.supported !== "boolean") {
    return [
      { value: "minimum", label: "Lowest available" },
      { value: "off", label: "Thinking off" },
    ];
  }
  // Even when the resolved result is Off, keep the Lowest available policy
  // selectable. Capability refresh must not replace a user's stored intent.
  if (cap.supported === false) return [
    { value: "minimum", label: "Lowest available" },
    { value: "off", label: "Thinking off" },
  ];

  // Provider default is deliberately not user-selectable. "minimum" is only
  // an alias for the first concrete capability option.
  const options = [{ value: "minimum", label: "Lowest available" }];
  const labels = {
    off: "Thinking off", on: "Thinking on",
    minimal: "Thinking minimal", low: "Thinking low", medium: "Thinking medium",
    high: "Thinking high", xhigh: "Thinking xhigh", max: "Thinking max", ultra: "Thinking ultra",
  };
  for (const value of concreteReasoningPreferences(cap))
    options.push({ value, label: labels[value] || `Thinking ${value}` });
  return options;
}

/** Resolve provider-neutral intent to one concrete, capability-proven mode. */
export function resolveReasoningPreference(value, reasoning) {
  const requested = normalizeReasoningPreference(value, "minimum");
  const cap = reasoning && typeof reasoning === "object" ? reasoning : {};
  const minimum = minimumReasoningPreference(cap);
  if (requested === "minimum") return minimum;
  // Off belongs to the user profile, not capability discovery. Unknown or
  // explicitly non-reasoning models keep Off. Only a model that explicitly
  // requires reasoning may clamp it to its verified minimum at dispatch time.
  if (requested === "off") {
    if (cap.mandatory === true) return minimum !== "default" ? minimum : "default";
    return "off";
  }
  const options = reasoningOptionsForCapability(cap);
  if (!options.length) return "default";
  if (options.some(option => option.value === requested)) return requested;
  return "default";
}

export function reasoningPreferenceIsActive(value, reasoning) {
  const selected = resolveReasoningPreference(value, reasoning);
  if (selected === "off") return false;
  if (selected === "on" || EFFORT_SET.has(selected)) return true;
  const cap = reasoning && typeof reasoning === "object" ? reasoning : {};
  return cap.mandatory === true || cap.default_enabled === true;
}
